lib/internal/loader/resolveRequestUrl.js

```diff
@@ -0,0 +1,129 @@
+'use strict';
+
+const {URL} = require('url');
+const internalCJSModule = require('internal/module');
+const NativeModule = require('native_module');
+const Path = require('path');
+
+const {ModuleWrap} = process.binding('module_wrap');
+
+const search = require('internal/loader/search');
+const asyncReadFile = require('util').promisify(require('fs').readFile);
+const debug = require('util').debuglog('esm');
+
+const ArrayJoin = Function.call.bind(Array.prototype.join);
+const ArrayMap = Function.call.bind(Array.prototype.map);
+
+const {Error} = global;
+
+const createDynamicModule = (exports/*: string[]*/, url = '', evaluate/*: function? */) => {
+  debug(`creating ESM facade for ${url} with exports: ${ArrayJoin(exports, ', ')}`);
+  const names = ArrayMap(exports, name => `${name}`);
+  // sanitized ESM for reflection purposes
+  let src = `export let executor;
+  ${ArrayJoin(ArrayMap(names, name => `export let $${name}`), ';\n')}
+  ;(() => [
+    fn => executor = fn,
+    { exports: { ${
+      ArrayJoin(ArrayMap(names, name => `${name}: {
+        get: () => $${name},
+        set: v => $${name} = v
+      }`), ',\n')
+    } } }
+  ]);
+  `;
+  const reflectiveModule = new ModuleWrap(src, `cjs-facade:${url}`);
+  reflectiveModule.instantiate();
+  const [setExecutor, reflect] = reflectiveModule.evaluate()();
+  // public exposed ESM
+  let reexports = `import { executor,
+    ${ArrayMap(names, name => `$${name}`)}
+  } from "";
+  export {
+    ${ArrayJoin(ArrayMap(names, name => `$${name} as ${name}`), ', ')}
+  }
+  // add await to this later if top level await comes along
+  typeof executor === "function" ? executor() : void 0;`;
+  if (typeof evaluate === 'function') {
+    setExecutor(() => evaluate(reflect));
+  }
+  const runner = new ModuleWrap(reexports, `${url}`);
+  runner.link(async () => reflectiveModule);
+  runner.instantiate();
+  return {
+    module: runner,
+    reflect
+  };
+}
+
+// Strategy for loading a standard JavaScript module
+class StandardModuleRequest {
+  constructor(url) {
+    this.url = url;
+  }
+
+  async createModule() {
+    const source = `${await asyncReadFile(this.url.pathname)}`;
```

Comment:
There's a [specialized function](https://github.com/nodejs/node/blob/master/lib/internal/url.js#L1371-L1379) you could use instead of `this.url.pathname`.